federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -707,29 +703,43 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
+			if util.ObjectMetaEquivalent(baseIngress.ObjectMeta, clusterIngress.ObjectMeta) &&
+				reflect.DeepEqual(baseIngress.Spec, clusterIngress.Spec) {
 				glog.V(4).Infof("Ingress %q in cluster %q does not need an update: cluster ingress is equivalent to federated ingress", ingress, cluster.Name)
 			} else {
 				glog.V(4).Infof("Ingress %s in cluster %s needs an update: cluster ingress %v is not equivalent to federated ingress %v", ingress, cluster.Name, clusterIngress, desiredIngress)
-				if !util.ObjectMetaEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) {
-					// Merge any annotations on the federated ingress onto the underlying cluster ingress,
-					// overwriting duplicates.
-					for key, val := range baseIngress.ObjectMeta.Annotations {
-						desiredIngress.ObjectMeta.Annotations[key] = val
-					}
-					ic.eventRecorder.Eventf(baseIngress, api.EventTypeNormal, "UpdateInCluster",
-						"Updating ingress in cluster %s", cluster.Name)
-
-					operations = append(operations, util.FederatedOperation{
-						Type:        util.OperationTypeUpdate,
-						Obj:         desiredIngress,
-						ClusterName: cluster.Name,
-					})
-					// TODO: Transfer any readonly (target-proxy, url-map etc) annotations from the master cluster to the federation, if this is the master cluster.
-					// This is only for consistency, so that the federation ingress metadata matches the underlying clusters.  It's not actually required.
+				objMeta, err := conversion.NewCloner().DeepCopy(clusterIngress.ObjectMeta)
+				if err != nil {
+					glog.Errorf("Error deep copying ObjectMeta: %v", err)
+				}
+				desiredIngress.ObjectMeta, ok = objMeta.(v1.ObjectMeta)
+				if !ok {
+					glog.Errorf("Internal error: Failed to cast to v1.ObjectMeta: %v", objMeta)
+				}
+				// Merge any annotations and labels on the federated ingress onto the underlying cluster ingress,
+				// overwriting duplicates.
+				if desiredIngress.ObjectMeta.Annotations == nil {
+					desiredIngress.ObjectMeta.Annotations = make(map[string]string)
+				}
+				for key, val := range baseIngress.ObjectMeta.Annotations {
+					desiredIngress.ObjectMeta.Annotations[key] = val
+				}
+				if desiredIngress.ObjectMeta.Labels == nil {
+					desiredIngress.ObjectMeta.Labels = make(map[string]string)
 				}
+				for key, val := range baseIngress.ObjectMeta.Labels {
+					desiredIngress.ObjectMeta.Labels[key] = val
+				}
+				ic.eventRecorder.Eventf(baseIngress, api.EventTypeNormal, "UpdateInCluster",
+					"Updating ingress in cluster %s", cluster.Name)
+
+				operations = append(operations, util.FederatedOperation{
+					Type:        util.OperationTypeUpdate,
+					Obj:         desiredIngress,
+					ClusterName: cluster.Name,
+				})
+				// TODO: Transfer any readonly (target-proxy, url-map etc) annotations from the master cluster to the federation, if this is the master cluster.
```

Comment:
The ingress controller in the cluster applies annotations to indicate what target-proxy, url-map etc it applied to this ingress.  It should all be consistent across all cluster ingresses, and should hence also be applied 'upwards' to the federated ingress for consistency.  Not 100% required, but ensures strict API compatibility between Kubernetes and Federation.  Make sense?